### PR TITLE
feat: Add name overwrite API

### PR DESCRIPTION
**Pull request by robertpin**
Also add breakout rooms updated event

### CODE REVIEW

File: modules/API/API.js [hunk 1]
```diff
@@ -428,6 +429,17 @@ function initCommands() {
                 logger.error('Failed sending endpoint text message', err);
             }
         },
+        'overwrite-name': (id, name) => {
+            logger.debug('Overwrite name command received');
+            let breakoutRoom = false, identifier = id;
+
+            if (id.indexOf('@') !== -1) {
```
Comment (saghul) [inline modules/API/API.js:436]: Can we hide this and take care of it inside `overwriteParticipantName` ? I think it makes for a cleaner API.

File: react/features/base/participants/reducer.js [hunk 2]
```diff
@@ -69,7 +70,8 @@ const DEFAULT_STATE = {
     sortedRemoteVirtualScreenshareParticipants: new Map(),
     sortedRemoteParticipants: new Map(),
     sortedRemoteScreenshares: new Map(),
-    speakersList: new Map()
+    speakersList: new Map(),
+    overwrittenNameList: {}
```
Comment (saghul) [inline react/features/base/participants/reducer.js:74]: style, alpha sort

File: react/features/breakout-rooms/middleware.js [hunk 3]
```diff
@@ -25,6 +26,7 @@ StateListenerRegistry.register(
 
             conference.on(JitsiConferenceEvents.BREAKOUT_ROOMS_UPDATED, ({ rooms, roomCounter }) => {
                 logger.debug('Room list updated');
+                APP.API.notifyBreakoutRoomsUpdated(rooms, roomCounter);
```
Comment (saghul) [inline react/features/breakout-rooms/middleware.js:29]: This will break mobile.

File: react/features/breakout-rooms/middleware.js [hunk 4]
```diff
@@ -36,16 +38,48 @@ StateListenerRegistry.register(
...
+
+        if (Object.keys(overwrittenNameList).length > 0) {
+            const newRooms = {};
+
+            Object.entries(action.rooms).forEach(([ key, r ]) => {
+                let participants = r?.participants || {};
+                let jid;
+
+                for (const id of Object.keys(overwrittenNameList)) {
+                    jid = Object.keys(participants).find(p => p.endsWith(id));
```
Comment (saghul) [inline react/features/breakout-rooms/middleware.js:55]: This looks dangerous. Let's extract the resource ID (part after the last / ) and make sure we compare that, so we are sure it fully matches.

**Review by saghul — CHANGES_REQUESTED** [review]
Left some comments, PTAL.

File: react/features/base/participants/actions.js [hunk 5]
```diff
@@ -660,3 +661,20 @@ export function localParticipantAudioLevelChanged(level) {
         level
     };
 }
+
+/**
+ * Overwrites the name of the participant with the given id.
+ *
+ * @param {string} id - Participant id;.
+ * @param {string} name - New participant name.
+ * @param {boolean} isBreakoutRoom - Whether the participant is in a breakout room.
```
Comment (saghul) [inline react/features/base/participants/actions.js:670]: This is no longer needed.

File: react/features/base/participants/middleware.js [hunk 6]
```diff
@@ -231,6 +234,57 @@ MiddlewareRegistry.register(store => next => action => {
     case PARTICIPANT_UPDATED:
         return _participantJoinedOrUpdated(store, next, action);
 
+    case OVERWRITE_PARTICIPANT_NAME: {
```
Comment (saghul) [inline react/features/base/participants/middleware.js:237]: Won't dispatching an unmodified `UPDATE_BREAKOUT_ROOMS` action trigger the name update in the other middleware? At least for the breakout rooms branch, that is.

File: react/features/base/participants/actionTypes.ts [hunk 7]
```diff
@@ -212,3 +212,14 @@ export const RAISE_HAND_UPDATED = 'RAISE_HAND_UPDATED';
...
  */
 export const LOCAL_PARTICIPANT_AUDIO_LEVEL_CHANGED = 'LOCAL_PARTICIPANT_AUDIO_LEVEL_CHANGED'
+
+/**
+ * The type of Redux action which overwrites the name of a participant.
+ * {
+ *     type: OVERWRITE_PARTICIPANT_NAME,
+ *     id: string,
+ *     name: string,
+ *     isBreakoutRoom: boolean
```
Comment (saghul) [inline react/features/base/participants/actionTypes.ts:222]: Drop

File: react/features/base/participants/middleware.js (hunk 6 above)
Comment (robertpin) [reply react/features/base/participants/middleware.js:237]: I'll add another param to the dispatch to make sure that it doesn't

**Review by saghul — DISMISSED** [review]
LGTM with a comment. Please don't land it yet.

**Comment by robertpin** [comment]
Jenkins test this please

File: react/features/base/participants/middleware.js [hunk 8]
```diff
@@ -231,6 +236,72 @@ MiddlewareRegistry.register(store => next => action => {
     case PARTICIPANT_UPDATED:
         return _participantJoinedOrUpdated(store, next, action);
 
+    case OVERWRITE_PARTICIPANTS_NAMES: {
+        const { participantList } = action;
+
+        if (!Array.isArray(participantList)) {
+            return;
```
Comment (saghul) [inline react/features/base/participants/middleware.js:243]: log error? silently not doing anything can be confusing...

File: modules/API/API.js [hunk 9]
```diff
@@ -428,6 +429,11 @@ function initCommands() {
                 logger.error('Failed sending endpoint text message', err);
             }
         },
+        'overwrite-name': participantList => {
```
Comment (saghul) [inline modules/API/API.js:432]:
```suggestion
        'overwrite-names': participantList => {
```

File: modules/API/API.js [hunk 10]
```diff
@@ -428,6 +429,11 @@ function initCommands() {
...
+            logger.debug('Overwrite name command received');
```
Comment (saghul) [inline modules/API/API.js:433]:
```suggestion
            logger.debug('Overwrite names command received');
```

File: modules/API/external/external_api.js [hunk 11]
```diff
@@ -46,6 +46,7 @@ const commands = {
     kickParticipant: 'kick-participant',
     muteEveryone: 'mute-everyone',
     overwriteConfig: 'overwrite-config',
+    overwriteName: 'overwrite-name',
```
Comment (saghul) [inline modules/API/external/external_api.js:49]:
```suggestion
    overwriteNames: 'overwrite-names',
```

**Review by saghul — COMMENTED** [review]
LGTM with a few minor nits, PTAL.

**Review by saghul — APPROVED** [review]

**Pull request merged**